ledger/src/ui/src/bootloader.c

```diff
@@ -0,0 +1,373 @@
+/**
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2021 RSK Labs Ltd
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include <stdbool.h>
+
+#include "apdu.h"
+#include "bolos_ux.h"
+#include "bolos_ux_common.h"
+#include "bootloader.h"
+#include "defs.h"
+#include "err.h"
+#include "communication.h"
+#include "unlock.h"
+
+// Attestation context shorthand
+#define attestation_ctx (G_bolos_ux_context.attestation)
+
+// Signer authorization context shorthand
+#define sigaut_ctx (G_bolos_ux_context.sigaut)
+
+// Onboard context shorthand
+#define onboard_ctx (G_bolos_ux_context.onboard)
+
+// Operation being currently executed
+static unsigned char curr_cmd;
+// autoexec signature app
+static char autoexec;
+
+/*
+ * Reset all reseteable operations, only if the given operation is starting.
+ *
+ * @arg[in] cmd operation code
+ */
+static void reset_if_starting(unsigned char cmd) {
+    // Reset only if starting new operation (cmd != curr_cmd).
+    // Otherwise we already reset when curr_cmd started.
+    if (cmd != curr_cmd) {
+        curr_cmd = cmd;
+        reset_attestation(&attestation_ctx);
+        reset_signer_authorization(&sigaut_ctx);
+        reset_onboard_ctx(&onboard_ctx);
+    }
+}
+
+// Check if we allow this version of the app to execute.
+static int is_app_version_allowed(application_t *app) {
```

Comment:
Looking now at the uses of this function, it probably doesn't make a lot of sense to have it. What do you think?

Reply:
I hadn't noticed that before, but since you mentioned maybe it makes sense to get rid of this function and use `is_authorized_signer` directly, what do you think?

Reply:
I agree, let's do that.